lessons/misc/intro-bash/lesson.md

```diff
@@ -0,0 +1,354 @@
+
+```
+grep 'fish' animals.txt
+```
+
+Notice that by default (and like most commands), `grep` prints its output to the shell window. Of course, this is not always what we want. In this case, we just want to count the occurences of *fish* rather that list them. In that case, we can *pipe* (`|`) the output of `grep` to the `wc` command (for **w**ord **c**ount) with the `-l` flag to count the number of lines from the `grep` output.
+
+```
+grep 'fish' animals.txt | wc -l
+```
+
+## Redirect
+
+What if we wanted to create a separate file with the list of fish, called `fish.txt`. Here, we can use `>` to *redirect* the output of `grep` to a file called `fish.txt`
+
+```
+grep 'fish' animals.txt > fish.txt
+```
+
+Notice nothing is printed to the shell but a new file with all of the fish has appeared. Cool!
+
+Importantly, using `>` and passing the output to a file that already exists will overwrite that file. It your just want to add to the end of the file, you can *append* using the `>>` operator. Let's append all animals that start with the letter 'E' to `fish.txt` and rename the file to `fish-plus-E.txt`
+
+```
+grep '^E' animals.txt >> fish.txt && mv fish.txt fish-plus-E.txt
+```
+
+Note the use of `&&`. This allows us to run multiple commands in a single line. In fact, there are a few such commands, which I'll list here:
+
+- `;` separates commands like they are on separate lines
+- `&&` does the next command if the previous one succeeds
+- `||` does the next command if the previous one fails
+- `&` runs the previous command in the "background"
+
+# Running programs from the command-line
+
+It can be quite useful to write entire scripts and run them in their entirety. This is particularly helpful if running scripts (e.g. in R, Python, etc) on a server or coputer cluster where Graphical User Interfaces (GUI's) are uncommon.
+
+Before starting, let's change into our `scripts` folder. Note, we could run the scripts from aany folder by specifying the full path to the file, but for now we'll change over.
+
+```
+cd ../scripts
+```
+
+We'll be executing the python script `hello_world.py`. First, let's make sure that we have Python and that the computer knows where to find it (more on this in a bit). `which` is a helpful command for this as it tells us which *executable file* is being run when we type a command in the shell.
+
+```
+which python
+```
+
+If successful, the shell will print the full path to the executable used in performing the command. Incidentally, these are the files that are executed when your using your computer for everyday use; they're execution is simply hidden because you're often using a GUI.
+
+Let's go ahead and run our script
+
+```
+python hello_world.py
+```
+
+We can similarly execute bash scripts in the same way.
+
+```
+bash print-numbers.sh
```

Comment:
Might be good to `less` into this script before running it and show that bash scripts are also just plain text files containing the same bash commands we might run in the command line itself - there are no other bells or whistles